File: src/main/golang/leetcode/editor/cn/[207]Course Schedule.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
func canFinish(numCourses int, prerequisites [][]int) bool {
	//dfs
	//var graph = make([][]int, numCourses)
	//for _, tuple := range prerequisites {
	//	graph[tuple[1]] = append(graph[tuple[1]], tuple[0])
	//}
	//visited := make([]bool, numCourses)
	//path := make([]bool, numCourses)
	//var dfs func(start int)
	//var hasCycle bool
	//dfs = func(start int) {
	//	if path[start] {
	//		hasCycle = true
	//	}
	//	if visited[start] || hasCycle {
	//		return
	//	}
	//	visited[start] = true
	//	path[start] = true
	//	for _, next := range graph[start] {
	//		dfs(next)
	//	}
	//	path[start] = false
	//}
	//for i := 0; i < numCourses; i++ {
	//	dfs(i)
	//}
	//return !hasCycle
	//topo traversal
	var queue []int
	var graph = make([][]int, numCourses)
	indegree := make([]int, numCourses)
	for _, tuple := range prerequisites {
		from, to := tuple[1], tuple[0]
		indegree[to] += 1
		graph[from] = append(graph[from], to)
	}
	for i := 0; i < numCourses; i++ {
		if indegree[i] == 0 {
			queue = append(queue, i)
		}
	}
	var cnt int
	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]
		cnt += 1
		for _, next := range graph[curr] {
			indegree[next] -= 1
			if indegree[next] == 0 {
				queue = append(queue, next)
			}
		}
	}
	return cnt == numCourses
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[207]Course Schedule.py
from typing import List


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # topo sort
        # 1. build graph
        graph = [[] for _ in range(numCourses)]
        in_degree = [0] * numCourses
        for u, v in prerequisites:
            graph[v].append(u)
            in_degree[u] += 1
        # 2. bfs
        queue = [i for i in range(numCourses) if in_degree[i] == 0]
        while queue:
            u = queue.pop(0)
            for v in graph[u]:
                in_degree[v] -= 1
                if in_degree[v] == 0:
                    queue.append(v)
        return sum(in_degree) == 0
# leetcode submit region end(Prohibit modification and deletion)
